content/blog/2017/05/2017-05-12-kubernetes-journey-on-azure.adoc

```diff
@@ -0,0 +1,397 @@
+        }
+      }
+      stage('Test'){
+        steps {
+          sh 'Run tests'
+        }
+      }
+      stage('Deploy'){
+        steps {
+          sh './kubectl apply -R true -f my_project'
+        }
+      }
+    }
+  }
+----
+
+As the devil is in the details, it was not as easy as it looked at first sight.
+We quickly faced several issues.
+
+=== Order matters
+
+Some resources needed to be deployed before others. A workaround was to use
+numbers as file names. But this adds extra logic at file name level, for
+example:
+
+[source]
+----
+project/00-nginx-ingress
+project/09-www.jenkins.io
+----
+
+=== Portability
+
+The deployment environments needed to be the same across development machines
+and Jenkins host. Even if it's a well known challenge, it was not easy to
+achieve.  The more the project grew, the more our scripts needed additional
+tools (`make`, `bats`, `jq` `gpg`, etc).  The more tools we used, the more
+issues appeared because of the different versions used.
+
+Another challenge that emerged when dealing with different environments was:
+how to manage environment's specific configurations (dev, prod, etc).
+
+Is it better to define different configuration files per environment? Perhaps,
+but this means code duplications, or using file templates that require more
+tools (`sed`, `jinja2`, `erb`). And more work, but less code to write and
+maintain.
+
+There isn't a golden rule we discovered, and the answer is probably in between.
+
+In any case, the good news is that `Jenkinsfile` provides an easy way to
+execute tasks from a docker image and an image can contain all the necessary
+tools in our environment, we can even use different docker images for each
+stage along the way.
+
+In the following example, I use the `my_env` Docker image. It contains all the
+tools needed to test, validate, and deploy changes.
+
+.Jenkinsfile
+[source,groovy]
+----
+pipeline{
+  agent {
+    docker{
+      image 'my_env:1.0'
+    }
+  }
+  options{
+    buildDiscarder(logRotator(numToKeepStr: '10'))
+    disableConcurrentBuilds()
+    timeout(time: 1, unit: 'HOURS')
+  }
+  triggers{
+    pollSCM('* * * * *')
+  }
+  stages{
+    stage('Init'){
+      steps{
+        // Init everything required to deploy our infra
+        sh 'make init'
+      }
+    }
+    stage('Test'){
+      steps{
+       // Run tests to validate changes
+       sh 'make test'
+      }
+    }
+    stage('Deploy'){
+      steps{
+       // Deploy changes in production
+       sh 'make deploy'
+      }
+    }
+  }
+  post{
+    always {
+      sh 'make notify'
+    }
+  }
+}
+----
+
+=== Secret credentials
+
+Managing secrets is a big subject and brings with it many different
+requirements which are very hard to fulfill.  For obvious reasons, we couldn't
+publish the credentials used within the infra project.  On the other side, we
+needed to keep track and share them. Particularly for the Jenkins node that
```

Comment:
"Particularly for the Jenkins node that deploys our cluster" Why is this a standalone sentence?